internal/pkg/deploy/cloudformation/stack/env.go

```diff
@@ -128,6 +134,10 @@ func (e *EnvStackConfig) Parameters() ([]*cloudformation.Parameter, error) {
 			ParameterKey:   aws.String(envParamAppDNSDelegationRoleKey),
 			ParameterValue: aws.String(e.dnsDelegationRole()),
 		},
+		{
+			ParameterKey:   aws.String(EnvParamServiceDiscoveryEndpoint),
+			ParameterValue: aws.String(fmt.Sprintf(fmtServiceDiscoveryEndpoint, e.in.Name, e.in.AppName)),
+		},
```

Comment:
Can you refresh my memory for what happens for `env upgrade`?

If I'm reading right now we will change the `ServiceDiscoveryEndpoint` param to `env.app.local` even for existing stacks or am I understanding wrong?

Reply:
Three possibilities here: 
1. The stack was created with an earlier version and doesn't have `ServiceDiscoveryEndpoint` as a parameter. We don't have a good way of adding new parameters to an env stack with env upgrade, but we *can* set the default value of the param in the new version's CF. Therefore, when `env upgrade` gets run on a stack that doesn't have the parameter, the default value is set to `app.local` and that gets used because we're can't provide a value in `env upgrade` for the new parameter, just "use previous value," which defaults to the default param value if there isn't a previous.
2. The stack was created with a new template version on a new version of Copilot. The `ServiceDiscoveryEndpoint` parameter is set to `env.app.local` at deploy time.
3. The stack is upgraded from the new template version. Again, `env upgrade` uses the `UsePreviousValue` setting when creating the new parameters, so the param value will be set to `env.app.local` and its default value will be `app.local`.

Reply:
Makes perfect sense thank you!